Bottom: 57c49bb54c28ffd41da0b641c52c54a95e397ab5
Top:    1efd1a63fce0a680ae423da3f98f3397659c6870
Author: Ian Kent <raven@themaw.net>
Date:   2013-05-02 14:37:41 +0800

autofs-5.0.7 - use MS_UNBINDABLE


---

diff --git a/daemon/direct.c b/daemon/direct.c
index 228a666..b52f52c 100644
--- a/daemon/direct.c
+++ b/daemon/direct.c
@@ -38,6 +38,8 @@
 #define INCLUDE_PENDING_FUNCTIONS
 #include "automount.h"
 
+extern unsigned long mountflags;
+
 /* Attribute to create detached thread */
 extern pthread_attr_t th_attr_detached;
 
@@ -426,6 +428,12 @@ int do_mount_autofs_direct(struct autofs_point *ap,
 		goto out_err;
 	}
 
+	ret = mount(NULL, me->key, NULL, MS_UNBINDABLE|MS_REC, NULL);
+	if (ret) {
+		warn(ap->logopt,
+		     "failed to set autofs mount unbindable path %s", me->key);
+	}
+
 	ret = stat(me->key, &st);
 	if (ret == -1) {
 		error(ap->logopt,
@@ -776,6 +784,13 @@ int mount_autofs_offset(struct autofs_point *ap, struct mapent *me, const char *
 		goto out_err;
 	}
 
+	ret = mount(NULL, mountpoint, NULL, MS_UNBINDABLE|MS_REC, NULL);
+	if (ret) {
+		warn(ap->logopt,
+		     "failed to set mount trigger %s unbindable at %s",
+		     me->key, mountpoint);
+	}
+
 	ret = stat(mountpoint, &st);
 	if (ret == -1) {
 		error(ap->logopt,
diff --git a/daemon/indirect.c b/daemon/indirect.c
index f4476dc..131f6f0 100644
--- a/daemon/indirect.c
+++ b/daemon/indirect.c
@@ -37,6 +37,8 @@
 #define INCLUDE_PENDING_FUNCTIONS
 #include "automount.h"
 
+extern unsigned long mountflags;
+
 /* Attribute to create detached thread */
 extern pthread_attr_t th_attr_detached;
 
@@ -153,6 +155,13 @@ static int do_mount_autofs_indirect(struct autofs_point *ap, const char *root)
 		goto out_rmdir;
 	}
 
+	ret = mount(NULL, root, NULL, MS_UNBINDABLE|MS_REC, NULL);
+	if (ret) {
+		warn(ap->logopt,
+		     "failed to set autofs mount unbindable path %s at %s",
+		     ap->path, root);
+	}
+
 	free(options);
 	options = NULL;
 
diff --git a/include/automount.h b/include/automount.h
index e72fa0d..8a78ea0 100644
--- a/include/automount.h
+++ b/include/automount.h
@@ -75,6 +75,10 @@ int load_autofs4_module(void);
 #define SMB_SUPER_MAGIC    0x0000517BL
 #define CIFS_MAGIC_NUMBER  0xFF534D42L
 
+#ifndef MS_UNBINDABLE
+#define MS_UNBINDABLE (1 << 17)
+#endif
+
 /* This sould be enough for at least 20 host aliases */
 #define HOST_ENT_BUF_SIZE	2048
